src/audio/copier.c

```diff
@@ -0,0 +1,666 @@
+	buffer_set_comp(buf, dev, dir);
+	comp_writeback(dev);
+	irq_local_enable(flags);
+
+	dcache_writeback_invalidate_region(buf, sizeof(*buf));
+}
+
+static struct comp_dev *create_dai(struct comp_ipc_config *config,
+				   struct ipc4_copier_module_cfg *copier,
+				   union ipc4_connector_node_id *node_id,
+				   struct pipeline *pipeline)
+{
+	struct sof_uuid id = {0xc2b00d27, 0xffbc, 0x4150, {0xa5, 0x1a, 0x24,
+				0x5c, 0x79, 0xc5, 0xe5, 0x4b}};
+	const struct comp_driver *drv;
+	struct ipc_config_dai dai;
+	struct comp_dev *dev;
+	int ret;
+
+	drv = ipc4_get_drv((uint8_t *)&id);
+	if (!drv)
+		return NULL;
+
+	config->type = SOF_COMP_DAI;
+
+	memset(&dai, 0, sizeof(dai));
+	dai.dai_index = node_id->f.v_index;
+	dai.direction = node_id->f.dma_type % 2;
+	dai.is_config_blob = true;
+
+	switch (node_id->f.dma_type) {
+	case ipc4_hda_link_output_class:
+	case ipc4_hda_link_input_class:
+		dai.type = SOF_DAI_INTEL_HDA;
+		dai.is_config_blob = true;
+		break;
+	case ipc4_i2s_link_output_class:
+	case ipc4_i2s_link_input_class:
+		dai.type = SOF_DAI_INTEL_SSP;
+		dai.is_config_blob = true;
+		break;
+	case ipc4_alh_link_output_class:
+	case ipc4_alh_link_input_class:
+		dai.type = SOF_DAI_INTEL_ALH;
+		dai.is_config_blob = true;
+		dai.dai_index -= IPC4_ALH_MAX_NUMBER_OF_GTW;
+		break;
+	case ipc4_dmic_link_input_class:
+		dai.type = SOF_DAI_INTEL_DMIC;
+		dai.is_config_blob = true;
+		break;
+	default:
+		return NULL;
+	}
+
+	dev = drv->ops.create(drv, config, &dai);
+	if (!dev)
+		return NULL;
+
+	if (dai.direction == SOF_IPC_STREAM_PLAYBACK)
+		pipeline->sink_comp = dev;
+	else
+		pipeline->source_comp = dev;
+
+	pipeline->sched_id = config->id;
+
+	list_init(&dev->bsource_list);
+	list_init(&dev->bsink_list);
+
+	ret = comp_dai_config(dev, &dai, copier);
+	if (ret < 0)
+		return NULL;
+
+	return dev;
+}
+
+static struct comp_dev *copier_new(const struct comp_driver *drv,
+				   struct comp_ipc_config *config,
+				   void *spec)
+{
+	struct ipc4_copier_module_cfg *copier = spec;
+	union ipc4_connector_node_id node_id;
+	struct ipc *ipc = ipc_get();
+	struct copier_data *cd;
+	struct comp_dev *dev;
+	size_t size, config_size;
+
+	comp_cl_dbg(&comp_copier, "copier_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	dcache_invalidate_region(spec, sizeof(*copier));
+
+	config->frame_fmt = (copier->base.audio_fmt.valid_bit_depth >> 3) - 2;
+	dev->ipc_config = *config;
+
+	config_size = copier->gtw_cfg.config_length * sizeof(uint32_t);
+	dcache_invalidate_region((char *)spec + sizeof(*copier), config_size);
+
+	size = sizeof(*cd);
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, size);
+	if (!cd)
+		goto error;
+
+	size = sizeof(*copier);
+	mailbox_hostbox_read(&cd->config, size, 0, size);
+	cd->out_fmt[0] = cd->config.base.audio_fmt;
+
+	comp_set_drvdata(dev, cd);
+
+	list_init(&dev->bsource_list);
+	list_init(&dev->bsink_list);
+
+	if (copier->gtw_cfg.node_id != IPC4_INVALID_NODE_ID) {
```

Comment:
a comment describing what we are doing here.

Reply:
sure